flytekit/extras/persistence/s3_awscli.py

```diff
@@ -0,0 +1,162 @@
+import logging
+import os as _os
+import re as _re
+import string as _string
+import time
+from shutil import which as shell_which
+from typing import Dict, List, Optional
+
+from flytekit.common.exceptions.user import FlyteUserException as _FlyteUserException
+from flytekit.configuration import aws as _aws_config
+from flytekit.core.data_persistence import DataPersistence, DataPersistencePlugins
+from flytekit.tools import subprocess as _subprocess
+
+
+def _update_cmd_config_and_execute(cmd: List[str]):
+    env = _os.environ.copy()
+
+    if _aws_config.ENABLE_DEBUG.get():
+        cmd.insert(1, "--debug")
+
+    if _aws_config.S3_ENDPOINT.get() is not None:
+        cmd.insert(1, _aws_config.S3_ENDPOINT.get())
+        cmd.insert(1, _aws_config.S3_ENDPOINT_ARG_NAME)
+
+    if _aws_config.S3_ACCESS_KEY_ID.get() is not None:
+        env[_aws_config.S3_ACCESS_KEY_ID_ENV_NAME] = _aws_config.S3_ACCESS_KEY_ID.get()
+
+    if _aws_config.S3_SECRET_ACCESS_KEY.get() is not None:
+        env[_aws_config.S3_SECRET_ACCESS_KEY_ENV_NAME] = _aws_config.S3_SECRET_ACCESS_KEY.get()
+
+    retry = 0
+    while True:
+        try:
+            return _subprocess.check_call(cmd, env=env)
+        except Exception as e:
+            logging.error(f"Exception when trying to execute {cmd}, reason: {str(e)}")
+            retry += 1
+            if retry > _aws_config.RETRIES.get():
+                raise
+            secs = _aws_config.BACKOFF_SECONDS.get()
+            logging.info(f"Sleeping before retrying again, after {secs} seconds")
+            time.sleep(secs)
+            logging.info("Retrying again")
+
+
+def _extra_args(extra_args: Dict[str, str]) -> List[str]:
+    cmd = []
+    if "ContentType" in extra_args:
+        cmd += ["--content-type", extra_args["ContentType"]]
+    if "ContentEncoding" in extra_args:
+        cmd += ["--content-encoding", extra_args["ContentEncoding"]]
+    if "ACL" in extra_args:
+        cmd += ["--acl", extra_args["ACL"]]
+    return cmd
+
+
+class S3Persistence(DataPersistence):
+    """
+    DataPersistence plugin for AWS S3 (and Minio). Use aws cli to manage the transfer. The binary needs to be installed
+    separately
+
+    .. prompt::
+
+       pip install awscli
+
+    """
+
+    PROTOCOL = "s3://"
+    _AWS_CLI = "aws"
+    _SHARD_CHARACTERS = [str(x) for x in range(10)] + list(_string.ascii_lowercase)
+
+    def __init__(self, default_prefix: Optional[str] = None):
+        super().__init__(name="awscli-s3", default_prefix=default_prefix)
+
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the AWS cli is present
+        """
+        if not shell_which(S3Persistence._AWS_CLI):
+            raise _FlyteUserException("AWS CLI not found! Please install it with `pip install awscli`.")
+
+    @staticmethod
+    def _split_s3_path_to_bucket_and_key(path: str) -> (str, str):
```

Comment:
To fix the syntax error in mypy/pyright:
```suggestion
    def _split_s3_path_to_bucket_and_key(path: str) -> Tuple[str, str]:
```

Reply:
Done